### ISSUE TITLE
Instant loading does not work with section anchors in Safari

### ISSUE THREAD

**Issue by ybizeul** [squidfunk/mkdocs-material#3797]
### Contribution guidelines

- [X] I've read the [contribution guidelines](https://github.com/squidfunk/mkdocs-material/blob/master/CONTRIBUTING.md) and wholeheartedly agree

### I've found a bug and checked that ...

- [X] ... the problem doesn't occur with the `mkdocs` or `readthedocs` themes
- [X] ... the problem persists when all overrides are removed, i.e. `custom_dir`, `extra_javascript` and `extra_css`
- [X] ... the documentation does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem

### Description

Beginning with version 7.0.0 anchors defined on headings do not work with Safari 15.4 macOS

### Expected behaviour

Clicking the the link on the TOC should jump to the section in the main content
Using a URL with # anchor should open the page at the location.

### Actual behaviour

Page stays on top

### Steps to reproduce

Run website and click on a H3 link on the right side

### Package versions

- Python: `Python 3.8.7`
- MkDocs: `mkdocs, version 1.1.2 from /usr/local/lib/python3.8/site-packages/mkdocs (Python 3.8)`
- Material: `7.0.0`

### Configuration

```yaml
site_name: Test Site
nav:
  - Home: /
theme:
  name: material
  features:
     - navigation.instant
markdown_extensions:
  - toc:
      permalink:
```

### System information

- Operating system: macOS 12.3.1
- Browser: Safari 15.4

**Comment by squidfunk** [comment]
Thanks for reporting. The `readthedocs` theme is probably an error in your configuration, right?

**Comment by ybizeul** [comment]
Ah yes I was testing with the readthedoc theme but couldn't make it to apply not sure why.

Fixing it

**Comment by squidfunk** [comment]
Thanks! Instant loading is a feature of Material for MkDocs. The `readthedocs` theme has no support for it.

**Comment by squidfunk** [comment]
I'm not able to reproduce what you are describing. Anchors within the page, as well as anchors between pages work for me:

https://user-images.githubusercontent.com/932156/161585160-3d28c277-3d44-49ff-badd-33e9f3e20df1.mp4

Please provide a screencast and explain exactly what you're observing. Either I'm not yet understanding the issue you're reporting, or it does not reproduce. Note that I only have Safari 15.2 on my hands, so it might be related to that.

**Comment by ybizeul** [comment]
https://user-images.githubusercontent.com/182035/161601783-22570f4a-7441-48df-a1ee-164fc067097b.mp4

**Comment by squidfunk** [comment]
Still not reproducible for me.

- Is there anything printed in the console?
- Could you please provide a self-contained example? You can just zip your project

**Comment by ybizeul** [comment]
You can see the console in the screencast, nothing special.
[mkdocs-materials-anchors 2.zip](https://github.com/squidfunk/mkdocs-material/files/8415887/mkdocs-materials-anchors.2.zip)

**Comment by squidfunk** [comment]
I meant the browser's console 😉

**Comment by ybizeul** [comment]
Ah sorry ! Nothing in the console

**Comment by ybizeul** [comment]
Were you able to reproduce the issue ?

**Comment by squidfunk** [comment]
Unfortunately not. I've tested with 7.0.0 and 8.2.8 (latest). Both are working fine. Here's a screencast with 8.2.8.:

https://user-images.githubusercontent.com/932156/162064632-da952d86-f3e0-4fef-bc20-f0d506b4431c.mp4

I'm afraid I can't put any more time into this issue, it just does not reproduce for me. I tried everything. If other users are having the same problem in the future, maybe we can find a way to create a reproduction, but this one just doesn't do it. Note that I'm running Safari 15.2 (not the latest 15.4), so it might be related to that, but until now it seems you're the only one with this problem. I'm closing this issue as not reproducible for now.

**Comment by ybizeul** [comment]
So you're not even trying to reload the page and see if it gets to the anchor ? That's what I explain in my screencast

It might be an issue with Safari 15.4 I guess.

**Comment by squidfunk** [comment]
Oh, sorry, I missed the reload. It wasn't apparent in your screencast and you did not mention it explicitly. I can reproduce it when reloading. Reopened. I'll look into it when I find some time!

**Comment by squidfunk** [comment]
So, a heads up: I'm aware of this issue, but solving it turns out to be very complex. I'm going to take this as an opportunity to refactor instant loading, which definitely needs to be simplified, also solving the issue at hand. However, I need to find enough time to do this, so I currently can't put an estimate on when this will be fixed.

In the meantime, disabling instant loading is a good workaround.

**Comment by ybizeul** [comment]
That’s what I did. Thanks for looking into this !

**Comment by squidfunk** [comment]
#5032 contains a refactoring of instant loading. I'm not sure that it's finished yet, but if you find some time for testing, that would be amazing. You can install it with:

```
pip install git+https://github.com/squidfunk/mkdocs-material.git@refactor/instant-loading
```

**Comment by feasgal** [comment]
I've had `navigation.instant` disabled for the last year due to the same issue. I tried the fix you linked yesterday. It seemed to solve the issue.

1. Enabled `navigation.instant` and checked to be sure the bug was still occurring in Safari.
    - It was. Anchor links did not work in Safari.

2. Uninstalled and reinstalled using the command line you supplied.
3. Checked to see if the anchor links now worked in Safari.
    - They did.

I didn't have time to thoroughly regression-test or anything.

**Comment by squidfunk** [comment]
The issues should be fixed in #5023. I've taken the time and did a complete rewrite of instant loading. It should be much more stable now. I'll issue a new release shortly, so we can find any remaining bugs quickly, but I'm quite confident that scroll restoration should work much better now. I've tested in Chrome, Firefox and Safari.

**Comment by squidfunk** [comment]
Released as part of [9.1.1](https://github.com/squidfunk/mkdocs-material/releases/tag/9.1.1).
